Command Mode Applications don't work with Kotlin 1.7

### Describe the bug

If you create a Command Mode Application for kotlin and change the kotlin version to 1.7 the application fails to work .


### Expected behavior

- When using `@QuarkusMain` the application should execute the `run()` method of the QuarkusApplication
- When using `quarkus.package.main-class` the application should execute the `run()` method of the QuarkusApplication
- When using `@Command` the application should execute the `run()` method of the Runnable 

### Actual behavior

- When using `@QuarkusMain` the application _**silently does not execute**_ the `run()` method of the QuarkusApplication
- When using `quarkus.package.main-class` the application fails to start with:
  ```
  ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#mainClassBuildStep threw an exception: java.lang.IllegalArgumentException: The supplied 'main-class' value of 'de.exaring.commandmode.MyMain' does not correspond to either a fully qualified class name or a matching 'name' field of one of the '@QuarkusMain' annotations
  ```
- When using `@Command` the application fails to start with:
  ```
  Exception in thread "main" java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "io.quarkus.deployment.dev.DevModeContext$ModuleInfo.getMain()" because the return value of "io.quarkus.deployment.dev.DevModeContext.getApplicationRoot()" is null
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:137)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
  Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkus.deployment.dev.DevModeContext$ModuleInfo.getMain()" because the return value of "io.quarkus.deployment.dev.DevModeContext.getApplicationRoot()" is null
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:90)
  ```

### How to Reproduce?

Steps to reproduce:
- run `quarkus create cli org.acme:kotlin-cli-test --extension=kotlin --gradle`
- edit the generated `build.gradle` and set kotlin-plugin versions to `1.7.0`
- run `quarkus dev`

### Output of `uname -a` or `ver`

Linux pop-os 5.17.5-76051705-generic

### Output of `java -version`

openjdk version "17.0.2" 2022-01-18

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.4.2

### Additional information

It works, if you downgrade kotlin to `1.6.21` in the `build.gradle` again

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -148,7 +148,7 @@
         <aws-lambda-java-events.version>3.11.0</aws-lambda-java-events.version>
         <aws-xray.version>2.11.2</aws-xray.version>
         <azure-functions-java-library.version>1.4.2</azure-functions-java-library.version>
-        <kotlin.version>1.6.21</kotlin.version>
+        <kotlin.version>1.7.10</kotlin.version>
         <kotlin.coroutine.version>1.6.4</kotlin.coroutine.version>
         <kotlin-serialization.version>1.3.3</kotlin-serialization.version>
         <kubernetes-client.version>5.12.3</kubernetes-client.version> <!-- Please check with Java Operator SDK team before updating -->

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -20,8 +20,8 @@
         <!-- These properties are needed in order for them to be resolvable by the generated projects -->
         <!-- Quarkus uses jboss-parent and it comes with 3.8.1-jboss-1, we don't want that in the templates -->
         <compiler-plugin.version>3.8.1</compiler-plugin.version>
-        <kotlin.version>1.6.21</kotlin.version>
-        <dokka.version>1.6.21</dokka.version>
+        <kotlin.version>1.7.10</kotlin.version>
+        <dokka.version>1.7.10</dokka.version>
         <scala.version>2.13.8</scala.version>
         <scala-maven-plugin.version>4.7.1</scala-maven-plugin.version>
         <!-- not pretty but this is used in the codestarts and we don't want to break compatibility -->

--- a/devtools/gradle/gradle-model/build.gradle
+++ b/devtools/gradle/gradle-model/build.gradle
@@ -1,5 +1,6 @@
 dependencies {
     implementation "io.quarkus:quarkus-core-deployment:${version}"
+    implementation "org.jetbrains.kotlin:kotlin-gradle-plugin-api:${kotlin_version}"
     testImplementation "io.quarkus:quarkus-devtools-testing:${version}"
 }
 

--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java
@@ -30,6 +30,7 @@ import org.gradle.api.tasks.SourceSetContainer;
 import org.gradle.api.tasks.compile.AbstractCompile;
 import org.gradle.language.jvm.tasks.ProcessResources;
 import org.gradle.tooling.provider.model.ParameterizedToolingModelBuilder;
+import org.jetbrains.kotlin.gradle.tasks.KotlinJvmCompile;
 
 import io.quarkus.bootstrap.BootstrapConstants;
 import io.quarkus.bootstrap.model.ApplicationModel;
@@ -436,6 +437,7 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
 
         final List<SourceDir> sourceDirs = new ArrayList<>(1);
         final List<SourceDir> resourceDirs = new ArrayList<>(1);
+
         project.getTasks().withType(AbstractCompile.class, t -> {
             if (!t.getEnabled()) {
                 return;
@@ -448,19 +450,45 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
             if (!allClassesDirs.contains(destDir)) {
                 return;
             }
-            final List<File> srcDirs = new ArrayList<>(1);
             source.visit(a -> {
                 // we are looking for the root dirs containing sources
                 if (a.getRelativePath().getSegments().length == 1) {
                     final File srcDir = a.getFile().getParentFile();
-                    if (srcDirs.add(srcDir)) {
+                    DefaultSourceDir sources = new DefaultSourceDir(srcDir.toPath(), destDir.toPath(),
+                            Collections.singletonMap("compiler", t.getName()));
+                    sourceDirs.add(sources);
+                }
+            });
+        });
+
+        // This "try/catch" is needed because of the way the "quarkus-cli" Gradle tests work. Without it, the tests fail.
+        try {
+            Class.forName("org.jetbrains.kotlin.gradle.tasks.KotlinJvmCompile");
+            project.getTasks().withType(KotlinJvmCompile.class, t -> {
+                if (!t.getEnabled()) {
+                    return;
+                }
+                final FileTree source = t.getSources().getAsFileTree();
+                if (source.isEmpty()) {
+                    return;
+                }
+                final File destDir = t.getDestinationDirectory().getAsFile().get();
+                if (!allClassesDirs.contains(destDir)) {
+                    return;
+                }
+                source.visit(a -> {
+                    // we are looking for the root dirs containing sources
+                    if (a.getRelativePath().getSegments().length == 1) {
+                        final File srcDir = a.getFile().getParentFile();
                         DefaultSourceDir sources = new DefaultSourceDir(srcDir.toPath(), destDir.toPath(),
                                 Collections.singletonMap("compiler", t.getName()));
                         sourceDirs.add(sources);
                     }
-                }
+                });
             });
-        });
+        } catch (ClassNotFoundException e) {
+            // ignore
+        }
 
         final File resourcesOutputDir = sourceSet.getOutput().getResourcesDir();
 
@@ -476,14 +504,11 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
                 return;
             }
             final Path destDir = t.getDestinationDir().toPath();
-            final List<File> srcDirs = new ArrayList<>(1);
             source.getAsFileTree().visit(a -> {
                 // we are looking for the root dirs containing sources
                 if (a.getRelativePath().getSegments().length == 1) {
                     final File srcDir = a.getFile().getParentFile();
-                    if (srcDirs.add(srcDir)) {
-                        resourceDirs.add(new DefaultSourceDir(srcDir.toPath(), destDir));
-                    }
+                    resourceDirs.add(new DefaultSourceDir(srcDir.toPath(), destDir));
                 }
             });
         });

--- a/devtools/gradle/gradle.properties
+++ b/devtools/gradle/gradle.properties
@@ -1,1 +1,2 @@
 version = 999-SNAPSHOT
+kotlin_version = 1.7.10

--- a/independent-projects/arc/tests/pom.xml
+++ b/independent-projects/arc/tests/pom.xml
@@ -48,7 +48,7 @@
         <dependency>
             <groupId>org.jetbrains.kotlin</groupId>
             <artifactId>kotlin-stdlib</artifactId>
-            <version>1.6.21</version>
+            <version>1.7.10</version>
             <scope>test</scope>
         </dependency>
 

